src/net/socket/udp_sock.c

```diff
@@ -28,79 +29,133 @@
 #include <stdlib.h>
 
 static const struct sock_proto_ops udp_sock_ops_struct;
-const struct sock_proto_ops *const udp_sock_ops
-		= &udp_sock_ops_struct;
+const struct sock_proto_ops *const udp_sock_ops = &udp_sock_ops_struct;
 
 EMBOX_NET_SOCK(AF_INET, SOCK_DGRAM, IPPROTO_UDP, 1,
 		udp_sock_ops_struct);
 
-static int udp_sendmsg(struct sock *sk, struct msghdr *msg, int flags) {
-	size_t total_len, data_len, actual_len, skb_len;
-	struct sk_buff *skb;
-	const struct sockaddr_in *to;
-	const struct sockaddr *sockaddr;
-	int i, ret;
-
-	assert(sk);
-	assert(sk->o_ops);
-	assert(sk->o_ops->make_pack);
-	assert(msg);
-	assert(msg->msg_iov);
-	assert(msg->msg_iov->iov_base);
-
-	data_len = 0;
-	for (i = 0; i < msg->msg_iovlen; i++) {
-		data_len += msg->msg_iov[i].iov_len;
-	}
+static int iov_msg_len(struct iovec *iov, int iovlen) {
+	int len = 0;
 
-	total_len = data_len + UDP_HEADER_SIZE;
+	for (int i = 0; i < iovlen; i++) {
+		len += iov[i].iov_len;
+	}
 
-	skb = NULL;
-	sockaddr = (const struct sockaddr *)msg->msg_name;
+	return len;
+}
 
-	if (msg->msg_name != NULL) {
-		to = (const struct sockaddr_in *)msg->msg_name;
-	} else {
-		to = (const struct sockaddr_in *)&to_inet_sock(sk)->dst_in;
-	}
+int iov_stream_make_queue(size_t len, size_t hdr_size, struct sk_buff_head *queue,
+		struct sock *sk, const struct sockaddr *sockaddr) {
+	int to_alloc = len;
 
-	skb_len = i = 0;
-	while (i < msg->msg_iovlen) {
-		data_len -= skb_len;
-		actual_len = data_len + UDP_HEADER_SIZE;
+	skb_queue_init(queue);
 
-		ret = sk->o_ops->make_pack(sk, sockaddr, &actual_len, &skb);
+	while (to_alloc > 0) {
+		struct sk_buff *skb = NULL;
+		size_t actual_len = to_alloc + hdr_size;
 
-		if (ret != 0) {
+		int ret = sk->o_ops->make_pack(sk, sockaddr, &actual_len, &skb);
+		if (ret < 0) {
+			skb_queue_purge(queue);
 			return ret;
 		}
+		skb_queue_push(queue, skb);
+		to_alloc -= actual_len - hdr_size;
+	}
+
+	return len;
+}
+
+int iov_dgram_make_queue(size_t len, size_t hdr_size, struct sk_buff_head *queue,
```

Comment:
Actually, yes, it supposed to. It's a difference between dgram and stream sockets, dgram supposed to have packet boundaries and 1 transmission goes in 1 packet (at IP level). 
I tried to unify interfaces of both socket types, so it should be easier to handle their output, the queue. Functions that will operate on queue will not be different in dgram and socket stream.
If you have some concerns about simillarity of functions, feel free to tell them